src/helpers/transport/bytearrstream.rs

```diff
@@ -38,26 +34,35 @@ impl ByteArrStream {
         }
     }
 
-    /// returns [`Bytes`] of `size_in_bytes` length from the buffered chunks. Returns [`None`] if
-    /// there are less than `size_in_bytes` bytes in the buffer
+    /// returns [`Bytes`] in multiples of `size_in_bytes` length from the buffered chunks. Returns
+    /// [`None`] if there are less than `size_in_bytes` bytes in the buffer
     fn pop_front(&mut self) -> Option<Bytes> {
         // not enough bytes buffered
         if self.buffered_size < self.size_in_bytes {
             None
-        } else if self.buffered[0].len() > self.size_in_bytes as usize {
-            // if the first buffer is large enough, simply split it in-place.
-            // This is O(1) operation, and should be true majority of the time, assuming large size
-            // buffers
-            let out_bytes = self.buffered[0].split_to(self.size_in_bytes as usize);
-            self.buffered_size -= self.size_in_bytes;
-            Some(out_bytes)
+        } else if self.buffered[0].len() >= self.size_in_bytes as usize {
+            // split off as much of the first buffer as can be aligned. This is O(1) operation,
+            // and should occur once per buffer, assuming buffers are typically larger than
+            // `size_in_bytes`
+            let buff_len = u32::try_from(self.buffered[0].len()).unwrap();
+            let num_aligned = buff_len / self.size_in_bytes;
+            let out_count = num_aligned * self.size_in_bytes;
+            self.buffered_size -= out_count;
+
+            // if buffer is exactly aligned with `size_in_bytes`, remove and return it; otherwise,
+            // just split as much as can be aligned.
+            if out_count == buff_len {
+                Some(self.buffered.pop_front().unwrap())
+            } else {
+                Some(self.buffered[0].split_to(usize::try_from(out_count).unwrap()))
+            }
         } else {
-            // First buffer is too small, so we need to take across buffers.
-            // Will require a memcopy of u8's across buffers in order to create 1 contiguous buffer
-            // to return.
-            // Removes expended buffers after taking
-
+            // first buffer is smaller than `size_in_bytes`, so will need to combine bytes across
+            // buffers. Will require a memcpy of u8's across buffers to create 1 contiguous aligned
+            // chunk to return. Should occur at most once per buffer, assuming buffers are typically
+            // larger than `size_in_bytes`
             let mut out_bytes = Vec::with_capacity(self.size_in_bytes as usize);
```

Comment:
i suspect that if first buffer is smaller than `size_in_bytes` we return exactly `size_in_bytes` here? Should we also prefer to return as many aligned bytes as possible?

Reply:
you've got it right.

what would be a way to return more aligned bytes here? if first buffer < size_in_bytes, we'll need to create a new `Vec` to copy first buffer + some amount of second buffer. are you saying the size of that `Vec` should be the length of the first buf + the max aligned length of the second buf? so we would effectively memcpy all data from the body?

Reply:
as we discussed, it's unclear whether changes to this would perform better or worse. we may revisit this at a later date if it becomes a bottleneck, but for now, will leave as-is